daemon/containerd/image.go

```diff
@@ -2,14 +2,223 @@ package containerd
+		ID:           string(ctrdimg.Target.Digest),
+		OS:           ociimage.OS,
+		Architecture: ociimage.Architecture,
+		Config: &containertypes.Config{
+			Entrypoint:   ociimage.Config.Entrypoint,
+			Env:          ociimage.Config.Env,
+			Cmd:          ociimage.Config.Cmd,
+			User:         ociimage.Config.User,
+			WorkingDir:   ociimage.Config.WorkingDir,
+			ExposedPorts: exposedPorts,
+			Volumes:      ociimage.Config.Volumes,
+			Labels:       ociimage.Config.Labels,
+			StopSignal:   ociimage.Config.StopSignal,
+			StopTimeout:  ociimage.Config.StopTimeout,
+			ArgsEscaped:  ociimage.Config.ArgsEscaped,
+			Shell:        ociimage.Config.Shell,
+			OnBuild:      ociimage.Config.OnBuild,
+		},
+	}
+	if ociimage.Config.Healthcheck != nil {
+		img.V1Image.Config.Healthcheck = &containertypes.HealthConfig{
+			Test:        ociimage.Config.Healthcheck.Test,
+			Interval:    ociimage.Config.Healthcheck.Interval,
+			Timeout:     ociimage.Config.Healthcheck.Timeout,
+			StartPeriod: ociimage.Config.Healthcheck.StartPeriod,
+			Retries:     ociimage.Config.Healthcheck.Retries,
+		}
+
+	}
+	img.RootFS = rootfs
+
+	return containerdImage, img, nil
+}
+
+// resolveImage searches for an image based on the given
+// reference or identifier. Returns the descriptor of
+// the image, which could be a manifest list, manifest, or config.
+func (i *ImageService) resolveImage(ctx context.Context, refOrID string, platform *ocispec.Platform) (img containerdimages.Image, err error) {
+	parsed, err := reference.ParseAnyReference(refOrID)
+	if err != nil {
+		return containerdimages.Image{}, errdefs.InvalidParameter(err)
+	}
+
+	is := i.client.ImageService()
+
+	digested, ok := parsed.(reference.Digested)
+	if ok {
+		imgs, err := is.List(ctx, "target.digest=="+digested.Digest().String())
+		if err != nil {
+			return containerdimages.Image{}, errors.Wrap(err, "failed to lookup digest")
+		}
+		if len(imgs) == 0 {
+			return containerdimages.Image{}, images.ErrImageDoesNotExist{Ref: parsed}
+		}
+
+		return imgs[0], nil
+	}
+
+	ref := reference.TagNameOnly(parsed.(reference.Named)).String()
+
+	// If the identifier could be a short ID, attempt to match
+	if truncatedID.MatchString(refOrID) {
+		filters := []string{
+			fmt.Sprintf("name==%q", ref), // Or it could just look like one.
+			"target.digest~=" + strconv.Quote(fmt.Sprintf(`sha256:^%s[0-9a-fA-F]{%d}$`, regexp.QuoteMeta(refOrID), 64-len(refOrID))),
+		}
+		imgs, err := is.List(ctx, filters...)
+		if err != nil {
+			return containerdimages.Image{}, err
+		}
+
+		if len(imgs) == 0 {
+			return containerdimages.Image{}, images.ErrImageDoesNotExist{Ref: parsed}
+		}
+		if len(imgs) > 1 {
+			digests := map[digest.Digest]struct{}{}
+			for _, img := range imgs {
+				if img.Name == ref {
+					return img, nil
+				}
+				digests[img.Target.Digest] = struct{}{}
+			}
+
+			if len(digests) > 1 {
+				return containerdimages.Image{}, errdefs.NotFound(errors.New("ambiguous reference"))
```

Comment:
Maybe `errdefs.Conflict` would be more suitable?

Reply:
Conflict (`409`) is mostly intended if there's a conflict with another request (e.g. update a container that is in the process of being deleted); that said; "it's hard", as those status codes weren't really designed with APIs in mind. Also see the discussion on https://github.com/moby/moby/issues/44685

This page has some flow-charts that have been quite useful; https://www.codetinkerer.com/2015/12/04/choosing-an-http-status-code.html